aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -180,53 +205,21 @@ Tensor& logcumsumexp_out(const Tensor& self, int64_t dim, Tensor& result) {
   return result;
 }
 
-Tensor _cumsum_cpu(const Tensor& self, int64_t dim) {
-  Tensor result = at::empty_like(self, MemoryFormat::Contiguous);
-  cumsum_stub(self.device().type(), result, self, dim);
-  return result;
-}
-
-Tensor& _cumsum_out_cpu(const Tensor& self, int64_t dim, Tensor& result) {
-  cumsum_stub(self.device().type(), result, self, dim);
-  return result;
-}
-
-Tensor cumsum(const Tensor& self, int64_t dim, c10::optional<ScalarType> dtype) {
-  auto result = [&]() {
-    NoNamesGuard guard;
-    return at::_cumsum(integer_upcast(self, dtype), dim);
-  }();
-  namedinference::propagate_names(result, self);
-  return result;
-}
-
-Tensor& cumsum_(Tensor& self, int64_t dim, c10::optional<ScalarType> dtype) {
-  TORCH_CHECK(
-          !dtype.has_value() || (self.scalar_type() == dtype.value()),
-          "provided dtype must match the dtype of self tensor in cumsum. Got ",
-          toString(self.scalar_type()),
-          " and ",
-          toString(dtype.value()),
-          ".");
-
-  return at::_cumsum_out(self, self, dim);
-}
-
-Tensor& cumsum_out(const Tensor& self, int64_t dim, c10::optional<ScalarType> dtype, Tensor& result) {
-  // result type is favored over dtype; check that they match if provided (NumPy doesn't check)
-  TORCH_CHECK(
-      !dtype.has_value() || (result.scalar_type() == dtype.value()),
-      "provided dtype must match dtype of result in cumsum. Got ",
-      toString(result.scalar_type()),
-      " and ",
-      toString(dtype.value()),
-      ".");
-  {
-    NoNamesGuard guard;
-    at::_cumsum_out(result, self.toType(result.scalar_type()), dim);
+TORCH_IMPL_FUNC(cumsum_out)
+(const Tensor& self,
+ int64_t dim,
+ c10::optional<ScalarType> dtype,
+ const Tensor& result) {
+  NoNamesGuard guard;
+  if (self.dim() == 0) {
+    result.fill_(self);
+  } else if (self.numel() == 0) {
+    result.zero_();
+  } else {
+    dim = maybe_wrap_dim(dim, self.dim());
+    auto& result_mut = const_cast<Tensor&>(result);
+    cumsum_stub(self.device().type(), result_mut, self.to(result.scalar_type()), dim);
```

Comment:
If possible, changing the type signature of `cumsum_stub` would be better, as these tend to be fairly localized changes.